### ISSUE TITLE
Universal object query function

### ISSUE THREAD

**Issue by zyr17** [LPSim/backend#92]
Instead of implementing query functions everywhere, gather them to one interface, so we can re-use it and decrease bugs.

The query function should be:
- take `self.position` and `match` as input.
- give a better interface to describe situation like guyu DSL. No need to be implement another DSL or support it, but should improve the readability and decrease complexity when calling (compared with current `check_position_valid` function).
- output object lists, and whether to support sorting, filterling?
- How to type hint, or impossible?

**Comment by xqm32** [comment]
Maybe we can make something like [Textual](https://github.com/Textualize/textual)'s DOM Queries[^1] for the query function.

[^1]: https://textual.textualize.io/guide/queries/

**Comment by zyr17** [comment]
> Maybe we can make something like [Textual](https://github.com/Textualize/textual)'s DOM Queries[1](#user-content-fn-1-53e0da1da517b473c6fd1af6bf2ecce0) for the query function.
> ## Footnotes
> 
>     1. https://textual.textualize.io/guide/queries/ [↩](#user-content-fnref-1-53e0da1da517b473c6fd1af6bf2ecce0)

Yes something like that, but with simpler grammars, as our target structures are fixed.

**Comment by zyr17** [comment]
### querying objects
`ObjectPosition.query` to return object lists and `ObjectPosition.query_one` to return one object or None. 
  - takes`Match` and `command_string` as arguments. command contains:
  - `both / our / opponent` to select player_table, or `self` to select this character (caller area must be character or character_status). must appear first.
  - `[active / prev / next] or [deck / hand / character / team_status]` for player_table, `[weapon / artifact / talent] or [status]` for character, to select an object (first group) or object lists (second group). Note with `both`, select one object may also receive two objects.
  - `[with key=value]` to filter current objects that `x.key == value`. All types will be converted to str to compare. `with` can be used multiple times.
  - `and` to make multiple queries, and their results are combined.

### position checking
new `check_position_valid` takes target `ObjectPosition`,  `Match` and `command_string` as input. target position and match is optional. and without match, related command string will raise error.
  - `source / target` to select one position, or `both` to compare between two positions
  - if one position selected: `[pidx=? / cidx=? / area=? / active=(true|false)]` to check if position fulfills the situation. for area names, they are case insensitive. can use multiple times, and all of them should pass.
  - if two position selected (i.e. `both`), `[(pidx|cidx|area|id)=(same|diff)]` to compare two positions are same or not. can use multiple times, and all of them should pass.
  - `and` to make multiple checks. All check should pass.
